### PR TITLE
Signal Service 1.1.5

### PR DESCRIPTION
**Pull request by malnick**
- Updates to tests to POST from signal to test cache server
- Updates for security enhancements 
- Updates to reporters to do asynchronous report queries and multiple, generic reporter types
- Hard fails if cluster-id is not present

### CODE REVIEW

**Comment by malnick** [comment]
@spahl This PR has the updates to signal and the integration test we discussed this past week. It now POSTs to the test server, which should add more stability to the testing cycle. 

This also satisfies https://mesosphere.atlassian.net/browse/DCOS-6436

CC'ing @pyronicide

**Comment by malnick** [comment]
@mellenburg @cmaloney If you could review the test code updates that would be super.

**Comment by malnick** [comment]
@cmaloney @spahl This is ready for review.

File: test_util/integration_test.py [hunk 1]
```diff
@@ -1290,75 +1292,105 @@ def wait_for_endpoint():
 
     r = requests.get(test_cache_url)
 
+    # Handy for diagnosing strange signal service test behavior, not sure if we should
```
Comment (malnick) [inline test_util/integration_test.py]: @cmaloney Your call here

File: test_util/integration_test.py [hunk 2]
```diff
@@ -1260,10 +1260,12 @@ def wait_for_endpoint():
...
 
-    test_cache_url = "http://{}:{}/signal_test_cache".format(service_points[0].host, service_points[0].port)
+    test_host = service_points[0].host
+    test_port = service_points[0].port
+    test_cache_url = "http://{}:{}/signal_test_cache".format(test_host, test_port)
+
     cmd = """
-data=`/opt/mesosphere/bin/dcos-signal -report-host leader.mesos -test 2> /dev/null`;
-curl -H 'Content-Type: application/json' -X POST -d "$data" {};
+/opt/mesosphere/bin/dcos-signal -master-url http://leader.mesos -test-url {}  2> /dev/null;
```
Comment (mellenburg) [inline test_util/integration_test.py]: I definitely like this more than saving stdout and posting with curl!

Comment (cmaloney) [inline test_util/integration_test.py]: no way for us to keep stdout and stderr separate here? Be nice to be able to log what came to stderr, because the signal service could be trying to tell us important. (And if we don't dump it to /dev/null, it _should_ show up in the mesos task log.)

File: test_util/integration_test.py [hunk 3]
```diff
@@ -1290,75 +1292,105 @@ def wait_for_endpoint():
...
-                'provider': 'onprem',
-                'source': 'cluster',
-                'clusterId': '',
-                'customerKey': '',
-                'environmentVersion': '',
-                'variant': 'open'}
-            }
+        'diagnostics': {
+            'event': 'health',
+            'anonymousId': 'open /var/lib/dcos/cluster-id: no such file or directory',
```
Comment (cmaloney) [inline test_util/integration_test.py]: This doesn't seem right... Probably shouldn't be erroring...

File: test_util/integration_test.py [hunk 4]
```diff
@@ -1290,75 +1292,105 @@ def wait_for_endpoint():
...
+    }
+
+    # Insert the generic property data which is the same between all signal tracks
+    for k, v in generic_properties.items():
+        exp_data['diagnostics']['properties'][k] = v
+        exp_data['cosmos']['properties'][k] = v
+        exp_data['mesos']['properties'][k] = v
 
+    # Insert all the diagnostics data programmatically
     master_units = [
```
Comment (cmaloney) [inline test_util/integration_test.py]: Please don't change this indentation here. Makes a lot of noise in the diff (Can change whitespace, just do it separately from when you're making semantic changes).

File: test_util/integration_test.py (hunk 3 above)
Comment (malnick) [inline test_util/integration_test.py]: This test runs on an agent, so the cluster-id does not exist. The test overwrote the error with an empty string before.

File: test_util/integration_test.py (hunk 2 above)
Comment (malnick) [inline test_util/integration_test.py]: It runs the POST to stdout if you execute it with -v, I've also left the print statement in the test code so we can see the return. However, since this is executed via marathon, getting the stdout from the service being executed is rather difficult.

File: test_util/integration_test.py (hunk 3 above)
Comment (cmaloney) [inline test_util/integration_test.py]: definitely the API should not be returning "no such file or directory", that will lead to problems. It should return json null to indicate "no user id" most likely, and have that be distinguishable from the "empty user id" (empty string).

Comment (cmaloney) [inline test_util/integration_test.py]: Best is probably, since we're triggering the signal service manually, pass the option to change the location of the cluster-id file to one of our choosing, so we can verify the file gets loaded up, data sent properly.

File: test_util/integration_test.py (hunk 2 above)
Comment (cmaloney) [inline test_util/integration_test.py]: Inside the test itself it's difficult to get, but if you go to the DC/OS UI and find the task which completed, then there is the whole stdout + stderr in the sandbox. Generally I really don't like dumping messages into a black hole

File: test_util/integration_test.py (hunk 3 above)
Comment (malnick) [inline test_util/integration_test.py]: We purposefully had the error propagate into the signal service like that so we can see what errored out, since there are a few type of errors we can have loading the cluster-id file, null could indicate any one of those errors and we want to surface the actual problem in segment. 

There isn't an option to change the location of the cluster-id currently, since this is a well established file, whose location does not change. However, if we did have such an option would you like another marathon app to make a fake cluster-id on the agent? What would the test look like in that case? And would that test actually test the signal service in the way we expect? 

My opinion is that the signal service should be tested on a master, and that it should return a not empty string for cluster-id as we expect, but we simply can't do that right now.

File: test_util/integration_test.py (hunk 2 above)
Comment (malnick) [inline test_util/integration_test.py]: I'm still not sure what you're asking here. Do you want me to not dump to dev null?

**Comment by malnick** [comment]
@cmaloney This is ready for another pass. I updated signal to hard fail on cluster-id not being present and added a fake cluster-id file to test against.

File: .gitignore [hunk 5]
```diff
@@ -1,4 +1,5 @@
 __pycache__
+ext/dcos_installer/packages/
```
Comment (cmaloney) [inline .gitignore]: why are we adding this?

At the very least seems unrelated. None of the stuff any of our tooling does ends up here.

File: test_util/integration_test.py [hunk 6]
```diff
@@ -1260,11 +1260,23 @@ def wait_for_endpoint():
...
 
-    test_cache_url = "http://{}:{}/signal_test_cache".format(service_points[0].host, service_points[0].port)
+    test_host = service_points[0].host
+    test_port = service_points[0].port
+    test_cache_url = "http://{}:{}/signal_test_cache".format(test_host, test_port)
+
     cmd = """
-data=`/opt/mesosphere/bin/dcos-signal -report-host leader.mesos -test 2> /dev/null`;
-curl -H 'Content-Type: application/json' -X POST -d "$data" {};
+cleanup() {
```
Comment (cmaloney) [inline test_util/integration_test.py]: This cleanup will never get hit. Note we're also running inside a mesos task sandbox here. I think echo `"test_id" > test-cluster-id && /opt/mesosphere/bin/dcos-sign -cluster-id-path $PATH/test-cluster-id` (and the rest) would work better.

File: test_util/integration_test.py [hunk 7]
```diff
@@ -1290,75 +1302,105 @@ def wait_for_endpoint():
...
+        'provider': 'onprem',
+        'source': 'cluster',
+        'clusterId': 'test-id',
+        'customerKey': '',
+        'environmentVersion': '',
+        'variant': 'open'
+    }
+
+    # Insert the generic property data which is the same between all signal tracks
+    for k, v in generic_properties.items():
```
Comment (cmaloney) [inline test_util/integration_test.py]: Why not:
`exp_data['diagnostics']['properties'] = copy(generic_properties)` for all of these? (The copy may not even be necessary.

If we're updating existing dictionaries, then do a `.update()` - https://docs.python.org/3/library/stdtypes.html#dict.update

File: test_util/integration_test.py (hunk 6 above)
Comment (malnick) [inline test_util/integration_test.py]: Line 1279?

Comment (malnick) [inline test_util/integration_test.py]: We should probably put it before the sleep()

File: .gitignore [hunk 8]
```diff
@@ -1,4 +1,5 @@
 __pycache__
+ext/dcos-installer/packages/
```
Comment (malnick) [inline .gitignore]: Gen makes that packages/ directory in the dcos-installer dir when I execute it locally in my dev env. @spahl has seen me accidentally push .tar.xz files that were placed in it in the past, and I almost did it again testing this PR. Would like to leave it here so we don't repeat it.

File: test_util/integration_test.py [hunk 9]
```diff
@@ -1260,12 +1260,19 @@ def wait_for_endpoint():
...
-    test_cache_url = "http://{}:{}/signal_test_cache".format(service_points[0].host, service_points[0].port)
+    test_host = service_points[0].host
+    test_port = service_points[0].port
+    test_cache_url = "http://{}:{}/signal_test_cache".format(test_host, test_port)
+
     cmd = """
-data=`/opt/mesosphere/bin/dcos-signal -report-host leader.mesos -test 2> /dev/null`;
-curl -H 'Content-Type: application/json' -X POST -d "$data" {};
+ID_PATH="/${PWD}/test-cluster-id"
+echo -n 'test-id' > $ID_PATH
```
Comment (cmaloney) [inline test_util/integration_test.py]: the dcos-cluster-id service actually puts a new line into the cluster-id file, the signal service should do a `.strip()` to remove whitespace around the cluster id

**Comment by malnick** [comment]
@cmaloney `-n` removed.

**Comment by malnick** [comment]
@cmaloney I've got a few more changes to make to the signal code here, please don't ship this tonight.
